scalding-core/src/main/scala/com/twitter/scalding/typed/cascading_backend/AsyncFlowDefRunner.scala

```diff
@@ -122,37 +121,36 @@ class AsyncFlowDefRunner extends Writer { self =>
      * because we don't want to take the side effect twice.
      */
     def addForce[T](c: Config,
-      m: Mode,
       init: TypedPipe[T],
       opt: TypedPipe[T],
       p: Future[TypedPipe[T]]): (State, Boolean) =
 
-      forcedPipes.get((c, m, opt)) match {
+      forcedPipes.get((c, opt)) match {
         case None =>
-          (copy(forcedPipes = forcedPipes + ((c, m, opt) -> p),
-            initToOpt = initToOpt + (init -> opt)), true)
+          (copy(forcedPipes = forcedPipes + ((c, opt) -> p),
+            initToOpt = initToOpt + ((c, init) -> opt)), true)
         case Some(_) =>
-          (copy(initToOpt = initToOpt + (init -> opt)), false)
+          (copy(initToOpt = initToOpt + ((c, init) -> opt)), false)
       }
 
-    def getForce[T](c: Config,
-      m: Mode,
-      init: TypedPipe[T]): Option[Future[TypedPipe[T]]] =
+    def getForce[T](c: Config, init: TypedPipe[T]): Option[Future[TypedPipe[T]]] =
 
-      initToOpt.get(init).map { opt =>
-        forcedPipes.get((c, m, opt)) match {
+      initToOpt.get((c, init)).map { opt =>
+        forcedPipes.get((c, opt)) match {
           case None =>
             sys.error(s"invariant violation: initToOpt mapping exists for $init, but no forcedPipe")
           case Some(p) => p
         }
       }
   }
 
-  private[this] var state: State = State(Map.empty, HMap.empty, HMap.empty)
+  private[this] var state: State = State(FilesToCleanUp.empty, HMap.empty, HMap.empty)
 
   private def updateState[S](fn: State => (State, S)): S =
     mutex.synchronized {
-      val (st1, s) = fn(state)
+      val s0 = state
+      val (st1, s) = fn(s0)
+      require(state eq s0, "updateState has recursively modified state, programming error")
```

Comment:
@fwbrasil note this pattern, like the case of FlowStateMap could fail if fn also calls update. We never do that, and this is a private method, but to be paranoid we check that state has not changed during the call to fn.